src/components/utils/src/timer.cc

```diff
@@ -29,190 +29,129 @@
-
-  result.it_value.tv_sec = miliseconds / DateTime::MILLISECONDS_IN_SECOND;
-  result.it_value.tv_nsec = (miliseconds % DateTime::MILLISECONDS_IN_SECOND) *
-                            DateTime::NANOSECONDS_IN_MILLISECOND;
-  result.it_interval.tv_sec = 0;
-  result.it_interval.tv_nsec = 0;
-
-  return result;
-}
-
-timer_t StartPosixTimer(timer::Timer& trackable,
-                        const timer::Milliseconds timeout) {
-  LOG4CXX_AUTO_TRACE(logger_);
-  timer_t internal_timer = NULL;
-
-  sigevent signal_event;
-  signal_event.sigev_notify = SIGEV_THREAD;
-  signal_event.sigev_notify_attributes = NULL;
-  signal_event.sigev_value.sival_ptr = static_cast<void*>(&trackable);
-  signal_event.sigev_notify_function = timer::HandlePosixTimer;
-
-  if (timer_create(CLOCK_REALTIME, &signal_event, &internal_timer) ==
-      kErrorCode) {
-    int error_code = errno;
-    LOG4CXX_FATAL(logger_,
-                  "Can`t create posix_timer. Error("
-                      << error_code << "): " << strerror(error_code));
-    return NULL;
-  }
-  const itimerspec itimer = MillisecondsToItimerspec(timeout);
-
-  if (timer_settime(internal_timer, 0, &itimer, NULL) == kErrorCode) {
-    int error_code = errno;
-    UNUSED(error_code);
-    LOG4CXX_FATAL(logger_,
-                  "Can`t set timeout to posix_timer. Error("
-                      << error_code << "): " << strerror(error_code));
-    return NULL;
-  }
-  return internal_timer;
-}
-
-bool StopPosixTimer(timer_t timer) {
-  LOG4CXX_AUTO_TRACE(logger_);
-  const int resultCode = timer_delete(timer);
-  if (kErrorCode == resultCode) {
-    int error_code = errno;
-    LOG4CXX_ERROR(logger_,
-                  "Can`t delete posix_timer. Error("
-                      << error_code << "): " << strerror(error_code));
-    return false;
-  }
-  return true;
-}
-}  // namespace
-
-timer::Timer::Timer(const std::string& name, const TimerTask* task_for_tracking)
+timer::Timer::Timer(const std::string& name, TimerTask* task)
     : name_(name)
-    , task_(task_for_tracking)
-    , repeatable_(false)
-    , timeout_ms_(0u)
-    , is_running_(false)
-    , timer_(NULL) {
+    , task_(task)
+    , task_lock_()
+    , delegate_(this)
+    , thread_(threads::CreateThread(name_.c_str(), &delegate_))
+    , single_shot_(true) {
   LOG4CXX_AUTO_TRACE(logger_);
   DCHECK(!name_.empty());
   DCHECK(task_);
+  DCHECK(thread_);
+  LOG4CXX_DEBUG(logger_, "Timer " << name_ << " has been created");
 }
 
 timer::Timer::~Timer() {
   LOG4CXX_AUTO_TRACE(logger_);
-  LOG4CXX_DEBUG(logger_, "Timer is to be destroyed " << name_);
   Stop();
+  DCHECK(thread_);
+  DeleteThread(thread_);
   sync_primitives::AutoLock auto_lock(task_lock_);
   DCHECK(task_);
   delete task_;
+  LOG4CXX_DEBUG(logger_, "Timer " << name_ << " has been destroyed");
 }
 
-void timer::Timer::Start(const Milliseconds timeout, const bool repeatable) {
+void timer::Timer::Start(const Milliseconds timeout, const bool single_shot) {
```

Comment:
@anosach-luxoft  I am not sure, but in my opinion it would be nice to extract methods to Interface :
Start
StartRepeatable
It is much more clear that some boolean flags.